plugins/ua_nodestore_binary_encoded.c

```diff
@@ -0,0 +1,1163 @@
+
+    node->references = (UA_NodeReferenceKind*) UA_calloc(referencesSize, sizeof(UA_NodeReferenceKind));
+    for (size_t i = 0; i < referencesSize; i++) {
+        UA_NodeId referenceTypeId;
+        retval |= UA_NodeId_decodeBinary(&encodedBin, &offset, &referenceTypeId);
+        UA_Boolean isInverse;
+        retval |= UA_Boolean_decodeBinary(&encodedBin, &offset, &isInverse);
+        memcpy(&node->references[i].referenceTypeId, &referenceTypeId, sizeof(UA_NodeId));
+        node->references[i].isInverse = isInverse;
+
+           size_t refTargetsSize;
+           retval |= UA_UInt64_decodeBinary(&encodedBin, &offset, (UA_UInt64 *)&refTargetsSize);
+           node->references[i].refTargetsSize = refTargetsSize;
+           node->references[i].refTargets = (UA_ReferenceTarget*)UA_calloc(node->references[i].refTargetsSize,
+                                                 sizeof(UA_ReferenceTarget));
+           for (size_t j = 0; j < refTargetsSize; j++) {
+               UA_UInt32 targetHash;
+            retval |= UA_UInt32_decodeBinary(&encodedBin, &offset, &targetHash);
+            UA_ExpandedNodeId target;
+            retval |= UA_ExpandedNodeId_decodeBinary(&encodedBin, &offset, &target);
+            node->references[i].refTargets[j].targetHash = targetHash;
+            memcpy(&node->references[i].refTargets[j].target, &target, sizeof(UA_ExpandedNodeId));
+           }
+    }
+
+    switch(nodeClass) {
+    case UA_NODECLASS_OBJECT:
+           retval |= objectNodeDecode(&encodedBin, &offset, (UA_ObjectNode*) node);
+        break;
+    case UA_NODECLASS_VARIABLE:
+        retval |= variableNodeDecode(&encodedBin, &offset, (UA_VariableNode*) node);
+        break;
+    case UA_NODECLASS_METHOD:
+           retval |= methodNodeDecode(&encodedBin, &offset, (UA_MethodNode*) node);
+        break;
+    case UA_NODECLASS_OBJECTTYPE:
+        retval |= objectTypeNodeDecode(&encodedBin, &offset, (UA_ObjectTypeNode*) node);
+        break;
+    case UA_NODECLASS_VARIABLETYPE:
+           retval |= variableTypeNodeDecode(&encodedBin, &offset, (UA_VariableTypeNode*) node);
+        break;
+    case UA_NODECLASS_REFERENCETYPE:
+          retval |= ReferenceTypeNodeDecode(&encodedBin, &offset, (UA_ReferenceTypeNode*) node);
+        break;
+    case UA_NODECLASS_DATATYPE:
+          retval |= dataTypeNodeDecode(&encodedBin, &offset, (UA_DataTypeNode*) node);
+        break;
+    case UA_NODECLASS_VIEW:
+          retval |= viewNodeDecode(&encodedBin, &offset, (UA_ViewNode*) node);
+        break;
+    default:
+        break;
+    }
+
+    if(retval != UA_STATUSCODE_GOOD) {
+           UA_LOG_ERROR(UA_Log_Stdout, UA_LOGCATEGORY_SERVER, "The decodeNode failed with error : %s",
+                                          UA_StatusCode_name(retval));
+    }
+
+    NodeEntry *entry = container_of(node, NodeEntry, nodeId);
+    ++entry->refCount;
+    entry->deleted = true; // remove the decoded node from the memory when release is called!
```

Comment:
You could just remove the `deleted` member of the entries.
I think you always want to delete when the refocunt is zero.
So you don't need that at all.